task_test.go

```diff
@@ -3207,6 +3207,112 @@ func TestReference(t *testing.T) {
 	}
 }
 
+func TestVarInheritance(t *testing.T) {
+	enableExperimentForTest(t, &experiments.EnvPrecedence, "1")
+	t.Parallel()
+
+	tests := []struct {
+		name string
+		want string
+		call string
+	}{
+		{
+			name: "shell",
+			want: "shell\nshell\n",
+		},
+		{
+			name: "entrypoint-global-dotenv",
+			want: "entrypoint-global-dotenv\nentrypoint-global-dotenv\n",
+		},
+		{
+			name: "entrypoint-global-vars",
+			want: "entrypoint-global-vars\nentrypoint-global-vars\n",
+		},
+		{
+			// We can't send env vars to a called task, so the env var is not overridden
+			name: "entrypoint-task-call-vars",
+			want: "entrypoint-task-call-vars\nentrypoint-global-vars\n",
+		},
+		{
+			// Dotenv doesn't set variables
+			name: "entrypoint-task-call-dotenv",
+			want: "entrypoint-task-call-vars\nentrypoint-task-call-dotenv\n",
+		},
+		{
+			name: "entrypoint-task-call-task-vars",
+			want: "entrypoint-task-call-task-vars\nentrypoint-task-call-task-vars\n",
+		},
+		{
+			// Dotenv doesn't set variables
+			name: "entrypoint-task-dotenv",
+			want: "entrypoint-global-vars\nentrypoint-task-dotenv\n",
+		},
+		{
+			name: "entrypoint-task-vars",
+			want: "entrypoint-task-vars\nentrypoint-task-vars\n",
+		},
+		// {
+		// 	// Dotenv not currently allowed in included taskfiles
+		// 	name: "included-global-dotenv",
+		// 	want: "included-global-dotenv\nincluded-global-dotenv\n",
+		// },
+		{
+			name: "included-global-vars",
+			want: "included-global-vars\nincluded-global-vars\n",
+			call: "included",
+		},
+		{
+			// We can't send env vars to a called task, so the env var is not overridden
+			name: "included-task-call-vars",
+			want: "included-task-call-vars\nincluded-global-vars\n",
+			call: "included",
+		},
+		{
+			// Dotenv doesn't set variables
+			// Dotenv not currently allowed in included taskfiles (but doesn't error in a task)
+			name: "included-task-call-dotenv",
+			want: "included-task-call-vars\nincluded-global-vars\n",
+			call: "included",
+		},
+		{
+			name: "included-task-call-task-vars",
+			want: "included-task-call-task-vars\nincluded-task-call-task-vars\n",
+			call: "included",
+		},
+		{
+			// Dotenv doesn't set variables
+			// Somehow dotenv is working here!
+			name: "included-task-dotenv",
+			want: "included-global-vars\nincluded-task-dotenv\n",
+			call: "included",
+		},
+		{
+			name: "included-task-vars",
+			want: "included-task-vars\nincluded-task-vars\n",
+			call: "included",
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			var buff bytes.Buffer
+			os.Setenv("VAR", "shell")
+			os.Setenv("ENV", "shell")
+			e := task.Executor{
+				Dir:    fmt.Sprintf("testdata/var_inheritance/%s", test.name),
```

Comment:
I wonder if it would make sense to scope by `v2`. We could soon just create a `v3` dir and keep both tests for a while.

```suggestion
				Dir:    fmt.Sprintf("testdata/var_inheritance/v2/%s", test.name),
```

Reply:
I'm not sure I'm following. These are testing v3 functionality are they not?

Reply:
I'm sorry. I meant v3 and v4. 😅 

It was just an idea, though.

Reply:
Ah understood. I don't have an issue with this. I will update and merge